[JSC] Support ARM64 Disassembly Comments.

justinmichaud opened the pull request:
#### 739d9f24705c7b993d6517b428309b95211064c5
<pre>
[JSC] Support ARM64 Disassembly Comments.
<a href="https://bugs.webkit.org/show_bug.cgi?id=243358">https://bugs.webkit.org/show_bug.cgi?id=243358</a>

Reviewed by Yusuke Suzuki.

Add support for ARM64 disassembler comments in jit code. We use a global map of code
address -&gt; comments, which is updated during linking. Comments do nothing when
dumpDisassembly=0.

Also, we opt-out some MacroAssembler files from unified builds to fix build
errors caused by a RegisterID symbol conflict that this patch accidentally exposes.

Here is an example of a comment:

```
comment(scratchGPR, &quot; = callFrame-&gt;callee-&gt;executableOrRareData&quot;);
emitGetFromCallFrameHeaderPtr(CallFrameSlot::callee, scratchGPR);
loadPtr(Address(scratchGPR, JSFunction::offsetOfExecutableOrRareData()), scratchGPR);
auto hasExecutable = branchTestPtr(Zero, scratchGPR, TrustedImm32(JSFunction::rareDataTag));
loadPtr(Address(scratchGPR, FunctionRareData::offsetOfExecutable() - JSFunction::rareDataTag), scratchGPR);
hasExecutable.link(this);
comment(scratchGPR, &quot; = (&quot;, scratchGPR, &quot;: Executable)-&gt;codeBlock&quot;);
loadPtr(Address(scratchGPR, FunctionExecutable::offsetOfCodeBlockFor(kind)), scratchGPR);
```

becomes

```
&lt;44&gt; 0x12e51c02c:    ldur     x1, [fp, #24]; %x1 = callFrame-&gt;callee-&gt;executableOrRareData
&lt;48&gt; 0x12e51c030:    ldur     x1, [x1, #24]
&lt;52&gt; 0x12e51c034:    tbz      x1, #0, 0x12e51c03c -&gt; &lt;60&gt;
&lt;56&gt; 0x12e51c038:    ldur     x1, [x1, #47]
&lt;60&gt; 0x12e51c03c:    ldur     x1, [x1, #96]; %x1 = (%x1: Executable)-&gt;codeBlock
```

* Source/JavaScriptCore/JavaScriptCore.xcodeproj/project.pbxproj:
* Source/JavaScriptCore/Sources.txt:
* Source/JavaScriptCore/assembler/AbstractMacroAssembler.h:
(JSC::AbstractMacroAssembler::comment):
(JSC::AbstractMacroAssembler::lineComment):
(JSC::AbstractMacroAssembler::commentImpl):
* Source/JavaScriptCore/assembler/Comments.cpp: Copied from Source/JavaScriptCore/disassembler/ARM64Disassembler.cpp.
(JSC::CommentRegistry::initialize):
(JSC::CommentRegistry::singleton):
* Source/JavaScriptCore/assembler/Comments.h: Added.
(JSC::CommentRegistry::WTF_RETURNS_LOCK):
(JSC::CommentRegistry::registerCodeRange):
(JSC::CommentRegistry::unregisterCodeRange):
(JSC::CommentRegistry::comment):
(JSC::CommentRegistry::orderedKey):
(JSC::CommentRegistry::orderedKeyInverse):
* Source/JavaScriptCore/assembler/LinkBuffer.cpp:
(JSC::LinkBuffer::linkCode):
(JSC::LinkBuffer::linkComments):
* Source/JavaScriptCore/assembler/LinkBuffer.h:
* Source/JavaScriptCore/disassembler/ARM64Disassembler.cpp:
(JSC::tryToDisassemble):
* Source/JavaScriptCore/jit/AssemblyHelpers.cpp:
(JSC::AssemblyHelpers::jitAssertCodeBlockOnCallFrameWithType):
(JSC::AssemblyHelpers::jitAssertCodeBlockMatchesCurrentCalleeCodeBlockOnCallFrame):
* Source/JavaScriptCore/jit/ExecutableAllocator.cpp:
(JSC::ExecutableMemoryHandle::~ExecutableMemoryHandle):
* Source/JavaScriptCore/runtime/InitializeThreading.cpp:
(JSC::initialize):

remove unified sources for macroasm

Canonical link: <a href="https://commits.webkit.org/253085@main">https://commits.webkit.org/253085@main</a>
</pre>

justinmichaud commented on Source/JavaScriptCore/assembler/Comments.h:
```diff
@@ -0,0 +1,117 @@
+
+    void registerCodeRange(void* start, void* end, CommentMap&& map)
+    {
+        // if (LIKELY(!Options::dumpDisassembly()))
```
This is just so EWS can check this code path. It will be commented out when I submit.

justinmichaud commented on Source/JavaScriptCore/assembler/Comments.h:
```diff
@@ -0,0 +1,117 @@
+
+    void unregisterCodeRange(void* start, void* end)
+    {
+        // if (LIKELY(!Options::dumpDisassembly()))
```
ditto

justinmichaud commented on Source/JavaScriptCore/jit/AssemblyHelpers.cpp:
```diff
@@ -227,14 +228,18 @@ void AssemblyHelpers::jitAssertCodeBlockMatchesCurrentCalleeCodeBlockOnCallFrame
     if (block.codeType() != FunctionCode)
         return;
     auto kind = block.isConstructor() ? CodeForConstruct : CodeForCall;
+    lineComment("jitAssertCodeBlockMatchesCurrentCalleeCodeBlockOnCallFrame with code block type: ", kind);
```
Note that this results in a nop being emitted iff dump disassembly is turned on

Constellation commented on Source/JavaScriptCore/assembler/Comments.h:
```diff
@@ -0,0 +1,117 @@
+    inline uintptr_t orderedKeyInverse(uintptr_t in) { return std::numeric_limits<uintptr_t>::max() - in; }
+
+    Lock m_lock;
+    std::map<uintptr_t, std::pair<uintptr_t, CommentMap>> m_comments WTF_GUARDED_BY_LOCK(m_lock);
```
Use `StdMap` instead of `std::map`, to use FastMalloc internally.

Constellation commented on Source/JavaScriptCore/runtime/InitializeThreading.cpp:
```diff
@@ -29,6 +29,7 @@
 #include "config.h"
 #include "InitializeThreading.h"
 
+#include "Comments.h"
```
Let's rename it to `AssemblyComments.h` since there are so many `Comments` concepts (JS comments? Comment node in DOM? etc.).

Constellation commented on Source/JavaScriptCore/assembler/Comments.h:
```diff
@@ -0,0 +1,117 @@
+
+namespace JSC {
+
+class CommentRegistry {
```
Let's rename it to `AssemblyCommentRegistry` since `JSC::CommentRegistry` looks like a JS comment registry.

Constellation commented on Source/JavaScriptCore/assembler/Comments.h:
```diff
@@ -0,0 +1,117 @@
+            RELEASE_ASSERT(thisStart < thisEnd);
+        }
+
+        m_comments.emplace(orderedKey(start), std::make_pair(newEnd, WTFMove(map)));
```
Use `std::pair { }` instead of `std::make_pair`.

Constellation commented on Source/JavaScriptCore/assembler/LinkBuffer.cpp:
```diff
@@ -470,6 +472,20 @@ void LinkBuffer::allocate(MacroAssembler& macroAssembler, JITCompilationEffort e
+        auto key = reinterpret_cast<uintptr_t>(commentLocation);
+        RELEASE_ASSERT(!map.contains(reinterpret_cast<uintptr_t>(commentLocation)), 
+            "You cannot print more than one comment on the same line.");
+        map.add(key, str);
```
str's thread can be different (disasm thread and compiler thread can be different). Let's do `str.isolatedCopy()`.

Constellation commented on Source/JavaScriptCore/assembler/Comments.h:
```diff
@@ -0,0 +1,117 @@
+        if (it2 == map.end())
+            return { };
+
+        return { it2->value };
```
If we copy the string here, then we should do `it2->value.isolatedCopy()` since if there are two threads ref and deref this same string, then we have a bad time for ref-counting.

Constellation commented on Source/JavaScriptCore/assembler/AbstractMacroAssembler.h:
```diff
@@ -1019,6 +1021,27 @@ class AbstractMacroAssembler : public AbstractMacroAssemblerBase {
+        s.print(values...);
+        commentImpl(s.toString());
+        m_assembler.nop();
+    }
```
I think we should avoid emitting nop since comment / no-comment changes the code generation. We would like to avoid getting different code for `dumpDisassembly`. For example, if we are using `offset` for the code for debugging, `dumpDisassembly` will break it since it can emit a nop, and it makes debugging rather hard. I think `comment` feature is enough.

Constellation commented on Source/JavaScriptCore/assembler/Comments.h:
```diff
@@ -0,0 +1,117 @@
+            RELEASE_ASSERT(newEnd <= thisStart
+                || thisEnd <= newStart);
+            RELEASE_ASSERT(thisStart < thisEnd);
+        }
```
Let's do it only when `#if ASSERT_ENABLED`. `--dumpDisassembly=1` is legit option in release build and we do not want to make it super slow.

Constellation commented on Source/JavaScriptCore/assembler/LinkBuffer.cpp:
```diff
@@ -470,6 +472,20 @@ void LinkBuffer::allocate(MacroAssembler& macroAssembler, JITCompilationEffort e
+        map.add(key, str);
+    }
+
+    CommentRegistry::singleton().registerCodeRange(m_executableMemory->start().untaggedPtr(), m_executableMemory->end().untaggedPtr(), WTFMove(map));
```
Let's skip this if we don't have any comments.

Constellation commented on Source/JavaScriptCore/assembler/Comments.h:
```diff
@@ -0,0 +1,117 @@
+        Locker locker { m_lock };
+
+        auto it = m_comments.find(orderedKey(start));
+        RELEASE_ASSERT(it != m_comments.end());
```
Let's allow the code destruction without having comments.
```cpp
if (it == m_comments.end())
    return;
```

Constellation reviewed (approved):
Super nice. r=me with `lineComment` change.

MenloDorian commented on Source/JavaScriptCore/Sources.txt:
```diff
@@ -46,17 +46,18 @@ API/OpaqueJSString.cpp
 assembler/MacroAssembler.cpp
-assembler/MacroAssemblerARM64.cpp
-assembler/MacroAssemblerARMv7.cpp
+assembler/MacroAssemblerARM64.cpp @no-unify
```
Is this necessary?  Ditto for the rest below?  Isn't it only MacroAssemblerARMv7 that has the aliasing issue?

justinmichaud replied:
Yes, but I believe they all have the same typedef. I haven't proven that this could cause an issue, but I think these files might be big enough anyway?

Commit:
[JSC] Support ARM64 Disassembly Comments.
https://bugs.webkit.org/show_bug.cgi?id=243358

Reviewed by Yusuke Suzuki.

Add support for ARM64 disassembler comments in jit code. We use a global map of code
address -> comments, which is updated during linking. Comments do nothing when
dumpDisassembly=0.

Also, we opt-out some MacroAssembler files from unified builds to fix build
errors caused by a RegisterID symbol conflict that this patch accidentally exposes.

Here is an example of a comment:

```
comment(scratchGPR, " = callFrame->callee->executableOrRareData");
emitGetFromCallFrameHeaderPtr(CallFrameSlot::callee, scratchGPR);
loadPtr(Address(scratchGPR, JSFunction::offsetOfExecutableOrRareData()), scratchGPR);
auto hasExecutable = branchTestPtr(Zero, scratchGPR, TrustedImm32(JSFunction::rareDataTag));
loadPtr(Address(scratchGPR, FunctionRareData::offsetOfExecutable() - JSFunction::rareDataTag), scratchGPR);
hasExecutable.link(this);
comment(scratchGPR, " = (", scratchGPR, ": Executable)->codeBlock");
loadPtr(Address(scratchGPR, FunctionExecutable::offsetOfCodeBlockFor(kind)), scratchGPR);
```

becomes

```
<44> 0x12e51c02c:    ldur     x1, [fp, #24]; %x1 = callFrame->callee->executableOrRareData
<48> 0x12e51c030:    ldur     x1, [x1, #24]
<52> 0x12e51c034:    tbz      x1, #0, 0x12e51c03c -> <60>
<56> 0x12e51c038:    ldur     x1, [x1, #47]
<60> 0x12e51c03c:    ldur     x1, [x1, #96]; %x1 = (%x1: Executable)->codeBlock
```

* Source/JavaScriptCore/JavaScriptCore.xcodeproj/project.pbxproj:
* Source/JavaScriptCore/Sources.txt:
* Source/JavaScriptCore/assembler/AbstractMacroAssembler.h:
(JSC::AbstractMacroAssembler::comment):
(JSC::AbstractMacroAssembler::lineComment):
(JSC::AbstractMacroAssembler::commentImpl):
* Source/JavaScriptCore/assembler/Comments.cpp: Copied from Source/JavaScriptCore/disassembler/ARM64Disassembler.cpp.
(JSC::CommentRegistry::initialize):
(JSC::CommentRegistry::singleton):
* Source/JavaScriptCore/assembler/Comments.h: Added.
(JSC::CommentRegistry::WTF_RETURNS_LOCK):
(JSC::CommentRegistry::registerCodeRange):
(JSC::CommentRegistry::unregisterCodeRange):
(JSC::CommentRegistry::comment):
(JSC::CommentRegistry::orderedKey):
(JSC::CommentRegistry::orderedKeyInverse):
* Source/JavaScriptCore/assembler/LinkBuffer.cpp:
(JSC::LinkBuffer::linkCode):
(JSC::LinkBuffer::linkComments):
* Source/JavaScriptCore/assembler/LinkBuffer.h:
* Source/JavaScriptCore/disassembler/ARM64Disassembler.cpp:
(JSC::tryToDisassemble):
* Source/JavaScriptCore/jit/AssemblyHelpers.cpp:
(JSC::AssemblyHelpers::jitAssertCodeBlockOnCallFrameWithType):
(JSC::AssemblyHelpers::jitAssertCodeBlockMatchesCurrentCalleeCodeBlockOnCallFrame):
* Source/JavaScriptCore/jit/ExecutableAllocator.cpp:
(JSC::ExecutableMemoryHandle::~ExecutableMemoryHandle):
* Source/JavaScriptCore/runtime/InitializeThreading.cpp:
(JSC::initialize):

remove unified sources for macroasm

Canonical link: https://commits.webkit.org/253085@main
